text/0028-ed25519-validation/0028-ed25519-validation.md

```diff
@@ -0,0 +1,79 @@
+
+Furthermore, it is important to note that the holder of the secret key can produce more than one valid distinct signature. Such transactions with the same essence but different signatures are considered as double spends by the consensus protocol and handled accordingly. While this does not pose a problem for the core protocol, it may be a problem for 2nd layer solutions, similar to how [transaction malleability in bitcoin presented an issue for the lightning network](https://en.bitcoinwiki.org/wiki/Transaction_Malleability#How_Does_Transaction_Malleability_Affect_The_Lightning_Network.3F).
+
+# Detailed Design
+
+In order to have consistent validation of Ed25519 signatures for all edge cases and throughout different implementations, this RFC proposes explicit validation criteria. These three criteria **must** be checked to evaluate whether a signature is valid.
+
+Using the notation and Ed25519 parameters as described in the RFC 8032, the criteria are defined as follows:
+
+1. Accept non-canonical encodings of A and R.
+2. Reject values for S that are greater or equal than L.
+3. Use the equation [8][S]B = [8]R + [8][k]A' for validation.
+
+In the following, we will explain each of these in more detail.
+
+## Decoding
+
+The Curve25519 is defined over the finite field of order p=2<sup>255</sup>−19. A curve point (x,y) is encoded into its compressed 32-byte representation, namely by the 255-bit encoding of the field element y followed by a single sign bit that is 1 for negative x and 0 otherwise. This approach provides a unique encoding for each valid point. However, there are two classes of edge cases representing non-canonical encodings of valid points:
+- encoding a y-coordinate as y + p
+- encoding a curve point (0,y) with the sign bit set to 1
+
+In contrast to the Section [Decoding](https://tools.ietf.org/html/rfc8032#section-5.1.3) of RFC 8032, it is _not_ required that the encodings of A and R are canonical. As long as the corresponding (x,y) is a valid curve point, any of such edge cases will be accepted.
+
+It is worth noting that due to allowing different encodings of the same point, one cannot check point equality by doing byte per byte comparisons.
+
+## Validation 
+
+The RFC 8032 mentions two alternative verification equations:
+1. [8][S]B = [8]R + [8][k]A'
+2. [S]B = R + [k]A'
+
+Each honestly generated signature following RFC 8032 satisfies the second, cofactor-less equation and thus, also the first equation. However, the opposite is not true: There are solutions only satisfying the first but not the latter.<br> This ambiguity in RFC 8032 has led to the current situation in which different implementations rely on different verification equations. 
+
+In order to be consistent with batched verification, the group equation [8][S]B = [8]R + [8][k]A' _must_ be used for validations instead of [S]B = R + [k]A'.
```

Comment:
The issue here is we want to accept points (public keys, nonces) with torsion components which have order 8L, but the arithmetic is usually done mod L. Torsion components can be removed with cofactor clearing in both single and batch verification.

Reply:
The point behind this paragraph is that even if you only have signatures fulfilling R = [S]B - [k]A, the (cofactor-less) batch validation - as described in the Ed25519 - can still fail. Thus, the batch validation must be done using [8] to be correct. This means that when we want to have the very obvious property, that _a (single) signature is valid w.r.t. batched verification (for n=1) if and only if it is valid w.r.t. non-batched verification_ we must also use the [8] in the non-batched validation to be consistent. I agree that this is (too) short in the paragraph above, so what would you suggest?

Reply:
I think the point here must be whether we want to accept points (public key and nonce) from the prime-order subgroup or from the whole group (ie. points with small torsion components). In the latter case the verification is simpler - no need to check for point order to be prime, especially for batch verification. Otherwise batch verification condition `O ?= [-\sum_i z_i s_i] B + \sum_i [z_i] R_i + \sum_i [z_i k_i] A_i` is plain wrong in some cases. So cofactor clearing allows simpler code without loss of security. I'd change the sentence as follows:

In order to accept points -- public keys and nonces -- with small torsion components and avoid additional checks without loss of security the following single signature and batch verification conditions must be used:

O ?= [8](-[S]B + R + [k]A)

O ?= [8]([-\sum_i z_i s_i] B + \sum_i [z_i] R_i + \sum_i [z_i k_i] A_i)

Reply:
I have completely rephrased that paragraph and included a reference to the "Taming the many EdDSAs" paper that shows why confactored verification is the only valid solution if batch validation should be supported.
What do you think?

Reply:
Looks good now.

Reply:
The three verification criteria are self-sufficient. However, later we mention batch verification with is undefined here, readers would have to look it up somewhere else, in the Taming paper, for example. How about we add batch verification equation `O ?= [8]([-\sum_i z_i s_i] B + \sum_i [z_i] R_i + \sum_i [z_i k_i] A_i)`? It's a bit ugly, but that would show the need for cofactor clearing (without it this equation doesn't work), and thus it would be easier to argue the use of "cofactored" single signature verification.

Additionally, we can mention that scalars (`k` -- the output of hash function, and values `-\sum_i z_i s_i`, `z_i`, `z_i k_i` in batch equation) can be reduced mod L (which is better than just unreduced integers or reduction mod 8L).